### DISCUSSION TITLE
Differentiability

### DISCUSSION THREAD

**Discussion by mtbrolly** [team-ocean/veros#722] [Q&A] [answered]
Really nice model! I'm curious about the potential for veros to be made differentiable. I notice this was hinted at in the blog post [here](https://dionhaefner.github.io/2021/04/higher-level-geophysical-modelling/]). Is it still on the to do list?

**Comment by julianmak** [comment]
Would like to know too, because I might have some interest to apply for some fellowship money for me to make something like that happen (whether it be with Veros, or some other Python model).

(Can I get a clarification on something, I think I am being stupid and completely misunderstanding what Veros does since about 2018. Veros is ported from Carsten's pyOM2, but Veros does **NOT** use that Fortran pyOM2 as the backend right, i.e. there is no weird `f2py` things calling the Fortran codes to run it through the Python interface? Because I think that's why I have not been looking at Veros because for all that time I have been thinking Veros has no easy hope of being differentiable [at least not that sensibly]...)

**Reply by dionhaefner** [reply]
>  Veros is ported from Carsten's pyOM2, but Veros does NOT use that Fortran pyOM2 as the backend right, i.e. there is no weird f2py things calling the Fortran codes to run it through the Python interface? 

Correct, Veros is a pure Python re-implementation of pyOM2.

**Comment by dionhaefner** [comment, marked as answer]
Veros is in fact differentiable in many configurations, but AD is generally untested and may provide unstable gradients, performance issues, or silently wrong results.

When selecting a differentiable linear solver (e.g. via `VEROS_LINEAR_SOLVER=scipy_jax`), AD works as expected over functions containing `Veros.run` (e.g. via `jax.grad`, `jax.jacfwd`, ...). That said, there are some big limitations:

- You can't perform I/O, get diagnostic output, or write checkpoints.
- There's no gradient checkpointing or other measures to make reverse-mode AD memory efficient.
- You can't use MPI, so you can only do runs on a single device (CPU or GPU).
- Gradients will likely be unstable when rolling out over many time steps.

:point_up: can all be fixed with additional engineering effort, but would rely on volunteers to make happen. In the meantime, you're more than welcome to experiment with applying AD to Veros (and report back!).
